addons/controllers/cpi/vspherecpiconfig_utils.go

```diff
@@ -23,14 +25,64 @@ import (
 	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	clusterapiutil "sigs.k8s.io/cluster-api/util"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	cutil "github.com/vmware-tanzu/tanzu-framework/addons/controllers/utils"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/config"
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
 	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
 	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addonconfigs/cpi/v1alpha1"
 )
 
+type CPIConfigAction func(client.Object)
+
+// forClusterMappedByCPIConfig performs the specified action for the cluster owns the CPIConfig
+// For a CPIConfig to be owned by a Cluster, it should
+//  1. in the same namespace as the cluster
+//  2. not a template CPIConfig, a.k.a. has annotation tkg.tanzu.vmware.com/template-config
+//  3. has the Cluster as its OwnerReference
+func forClusterMappedByCPIConfig(cpiConfig client.Object, cluster *clusterapiv1beta1.Cluster,
+	controllerConfig config.ConfigControllerConfig, action CPIConfigAction) {
+
+	namespace := cpiConfig.GetNamespace()
+	annotations := cpiConfig.GetAnnotations()
+	ownerReferences := cpiConfig.GetOwnerReferences()
+
+	if namespace == cluster.Namespace {
+		// avoid enqueuing reconcile requests for template CPIConfig CRs in event handler of Cluster CR
+		if _, ok := annotations[constants.TKGAnnotationTemplateConfig]; ok && cpiConfig.GetNamespace() == controllerConfig.SystemNamespace {
+			return
+		}
+	}
+
+	// corresponding CPIConfig should have following ownerRef
+	ownerReference := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if clusterapiutil.HasOwnerRef(ownerReferences, ownerReference) {
```

Comment:
Should do the same for oracle cpi config?

Reply:
`forClusterMappedByCPIConfig` is used by both vsphere & oracle cpi config. we are finding all cpiconfigs that has owner reference to a cluster.